source/plugins/go/input/lib/cadvisor.go

```diff
@@ -0,0 +1,1219 @@
+
+	var defaultHost string
+	if cAdvisorSecurePort {
+		defaultHost = fmt.Sprintf("https://localhost:%s", CADVISOR_SECURE_PORT)
+	} else {
+		defaultHost = fmt.Sprintf("http://localhost:%s", CADVISOR_NON_SECURE_PORT)
+	}
+
+	var nodeIP string
+	if winNode != nil {
+		nodeIP = winNode["InternalIP"]
+	} else {
+		nodeIP = os.Getenv("NODE_IP")
+	}
+
+	if nodeIP != "" {
+		Log.Infof("Using %s for CAdvisor Host", nodeIP)
+		if cAdvisorSecurePort {
+			return fmt.Sprintf("https://%s:%s", nodeIP, CADVISOR_SECURE_PORT)
+		} else {
+			return fmt.Sprintf("http://%s:%s", nodeIP, CADVISOR_NON_SECURE_PORT)
+		}
+	} else {
+		Log.Warnf("NODE_IP environment variable not set. Using default as: %s", defaultHost)
+		if winNode != nil {
+			return ""
+		} else {
+			return defaultHost
+		}
+	}
+}
+
+func getCAdvisorUri(winNode map[string]string, relativeUri string) string {
+	baseCAdvisorUrl := getBaseCAdvisorUrl(winNode)
+	if baseCAdvisorUrl == "" {
+		return ""
+	}
+	return fmt.Sprintf("%s%s", baseCAdvisorUrl, relativeUri)
+}
+
+func isCAdvisorOnSecurePort() bool {
+	cAdvisorSecurePort := false
+	if os.Getenv("IS_SECURE_CADVISOR_PORT") == "true" {
+		cAdvisorSecurePort = true
+	}
+	return cAdvisorSecurePort
+}
+
+func getResponse(winNode map[string]string, relativeUri string) (*http.Response, error) {
+	var response *http.Response
+	Log.Infof("Getting CAdvisor Uri Response")
+	bearerToken, _ := ioutil.ReadFile(BEARER_TOKEN_LOCATION)
+
+	cAdvisorUri := getCAdvisorUri(winNode, relativeUri)
+	Log.Infof("CAdvisor Uri: %s", cAdvisorUri)
+
+	if cAdvisorUri != "" {
+		uri, err := url.Parse(cAdvisorUri)
+		if err != nil {
+			Log.Errorf("Failed to parse CAdvisor Uri: %s", err)
+			return nil, err
+		}
+
+		var httpClient *http.Client
+		if isCAdvisorOnSecurePort() {
+			caCert, err := ioutil.ReadFile(CERT_LOCATION)
+			if err != nil {
+				Log.Errorf("Failed to read CA cert: %s", err)
+				return nil, err
+			}
+
+			caCertPool := x509.NewCertPool()
+			caCertPool.AppendCertsFromPEM(caCert)
+
+			httpClient = &http.Client{
+				Transport: &http.Transport{
+					Proxy: http.ProxyFromEnvironment,
+					DialContext: (&net.Dialer{
+						Timeout:   20 * time.Second,
+						KeepAlive: 0,
+					}).DialContext,
+					TLSHandshakeTimeout:   20 * time.Second,
+					ExpectContinueTimeout: 1 * time.Second,
+					TLSClientConfig: &tls.Config{
+						RootCAs:            caCertPool,
+						InsecureSkipVerify: true,
+					},
+				},
+				Timeout: 40 * time.Second,
+			}
+		} else {
+			httpClient = &http.Client{
+				Timeout: 40 * time.Second,
+			}
+		}
+
+		req, err := http.NewRequest("GET", uri.String(), nil)
+		if err != nil {
+			Log.Errorf("Failed to create HTTP request: %s", err)
+			return nil, err
+		}
+		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", bearerToken))
+		response, err = httpClient.Do(req)
+		if err != nil {
+			Log.Errorf("Failed to get HTTP response: %s", err)
+			return nil, err
+		}
+		// defer response.Body.Close()
```

Comment:
since we are not closing response.Body.Close() and this can cause resource leak, and I understand, the reason why this is being commented out since the response being returned. I think, we would need to address this.

Reply:
arent we closing it below in the else clause "defer cAdvisorStats.Body.Close()"